src/net_processing.cpp

```diff
@@ -230,9 +232,7 @@ namespace {
             return &(*a) < &(*b);
         }
     };
-    std::map<COutPoint, std::set<std::map<uint256, COrphanTx>::iterator, IteratorComparator>> mapOrphanTransactionsByPrev GUARDED_BY(g_cs_orphans);
-
-    std::vector<std::map<uint256, COrphanTx>::iterator> g_orphan_list GUARDED_BY(g_cs_orphans); //! For random eviction
+    std::map<COutPoint, std::set<OrphanTxPool::iterator, IteratorComparator>> mapOrphanTransactionsByPrev GUARDED_BY(g_cs_orphans);
```

Comment:
`IteratorComparator` compares the addresses of the elements in the container! With `std::unordered_map`, I guess, if rehashing occurs, those could change in an unpredictable way, confusing the `std::set` container. Maybe at some point the `std::set` could end up having element `A` before element `B` (two iterators to the `std::unordered_map`) but they now compare `A > B` (after a rehash).

Reply:
This can be overcome by changing the `std::set` to `std::unordered_set`, removing the bizarre "order by address" `IteratorComparator` and hashing by the tx id. I checked that nowhere in the code we rely on the order of the elements in the `std::set` (as expected, because they are ordered by their address, which is "random" or unpredictable to the application). Something like this:

<details>
<summary>patch</summary>

```diff
diff --git i/src/net_processing.cpp w/src/net_processing.cpp
index d72edae6d..a02945d7e 100644
--- i/src/net_processing.cpp
+++ w/src/net_processing.cpp
@@ -221,21 +221,25 @@ namespace {
     /** Relay map */
     typedef std::map<uint256, CTransactionRef> MapRelay;
     MapRelay mapRelay GUARDED_BY(cs_main);
     /** Expiration-time ordered list of (expire time, relay map entry) pairs. */
     std::deque<std::pair<int64_t, MapRelay::iterator>> vRelayExpiration GUARDED_BY(cs_main);
 
-    struct IteratorComparator
+    class IteratorHash
     {
-        template<typename I>
-        bool operator()(const I& a, const I& b) const
+    public:
+        size_t operator()(const OrphanTxPool::iterator& it) const
         {
-            return &(*a) < &(*b);
+            return m_hasher(it->first);
         }
+
+    private:
+        const uint256Hash m_hasher;
     };
-    std::map<COutPoint, std::set<OrphanTxPool::iterator, IteratorComparator>> mapOrphanTransactionsByPrev GUARDED_BY(g_cs_orphans);
+
+    std::map<COutPoint, std::unordered_set<OrphanTxPool::iterator, IteratorHash>> mapOrphanTransactionsByPrev GUARDED_BY(g_cs_orphans);
 
     static size_t vExtraTxnForCompactIt GUARDED_BY(g_cs_orphans) = 0;
     static std::vector<std::pair<uint256, CTransactionRef>> vExtraTxnForCompact GUARDED_BY(g_cs_orphans);
 } // namespace
 
 namespace {
```
</details>

Reply:
Actually, there is a bigger problem which the above patch would not solve - all iterators to `mapOrphanTransactions` might be invalidated by an insertion to it:

https://github.com/bitcoin/bitcoin/blob/4286e3907d79199bc44e4cf6f1c45635339614e2/src/net_processing.cpp#L924

I.e. the above line might invalidate all iterators to `mapOrphanTransactions` that are stored in `mapOrphanTransactionsByPrev`.

Reply:
[Updated](https://github.com/bitcoin/bitcoin/pull/19374#issuecomment-650557986).